### ISSUE TITLE
chore: Move git-related stuff to its own module

### ISSUE THREAD

**Issue by spenserblack** [spenserblack/gengo#165]
For readability, and also makes #149 easier to implement in the future.

**Comment by spenserblack** [comment]
@Byron as the champion of gengo's parallelization, I'll include you in this :slightly_smiling_face: 

You don't need to make a PR, but I'd like to bounce some ideas off you, if possible.

Originally (draft in #150) my idea was that "file sources" would return an iterator `impl Iterator<Item=(Path, &[u8])>`. However, I don't think that's reasonable with the current implementation of parallelization. So I'm currently thinking, instead of "file sources", there would instead be "analyzer handlers." Vaguely like this:

```rust
// rough draft code
impl FileHandler {
    fn analyze<F: Fn(Path, &[u8]) -> Language>(f: F) {
        my_parallizer(|| {
            // f would be the analyzer function that identifies a file's language
            let results = f();
            // mutate gengo instance here?
        })
    }
}
```

What do you think? Am I going in a reasonable direction?

My goal is to make this as generic as possible, so that non-gix sources can be used, too. This `FileHandler` will probably also need to implement methods for returning overrides -- to get _really_ generic, and taking into account the possiblity of non-git sources, we might not want to assume the availability of `.gitattributes`.

Part of my inspiration is that I'm considering renaming the `gengo` exe to `git-gengo`, and making a `gengo` exe that reads from the work tree like tokei does.

**Comment by Byron** [comment]
> @Byron as the champion of gengo's parallelization, I'll include you in this 🙂

Hehe, parallelization in Rust is truly fearless, and I am sure you will consider yourself a champion soon, too!

I would avoid abstracting too much, but believe that `.gitattributes` should be reasonably abstracted away. However, I'd love to see the combination of `read files from worktree, but use `.gitattributes` with `Source::WorktreeThenIndex`, as well.

As for abstraction, I'd probably adjust `GitState` to also allow to have no repo, and pre-set the desired values for attributes in it. With that it should be possible to deal with combinations like "git-only", "worktree + gitattributes" and "any-directory, no git".

#### Parallelization

I'd go with walkdir + threads for doing the actual work, i.e. read-file + analyse. There is no need for parallel directory tree iteration as the bottleneck will be reading them. With this simple system, you will be maxing IO on the machine that runs it.

#### About Binaries

Personally I'd favor a `gengo` can do it all situation, but wouldn't mind if it could detect if it's run as `git gengo` to maybe have different defaults. To me it's definitely a 'finishing touch' though.


Hope that helps!

**Comment by spenserblack** [comment]
Got it, thanks for the perspective!

Just to clarify, `gix::parallel` isn't *required* in gengo, right? It's an abstraction to make parallelization simpler? It's an awesome tool, but IMO it sends the signal to a code reader "this tool is for gix," so if I write a more generic implementation I might drop it. Just an FYI that my style is often readability over performance -- I follow the philosophy that it's easier to make readable code performant than performant code readable :slightly_smiling_face:

So I think I'm reverting back to my older plan -- a `FileSource` trait that requires a method that returns `impl Iter<Item = (PathLikeType, &[u8])>`. And it would also provide `documentation_override(path, content) -> Option<bool>`, `generated_override`, etc. methods that would return `None` by default.

**Comment by Byron** [comment]
> Just to clarify, `gix::parallel` isn't _required_ in gengo, right? It's an abstraction to make parallelization simpler?

Yes, that's correct. You could see how it works and implement your own, or do parallelization in any way you see fit. It's something anyone can use, it's general and nothing specific to `gix`, but I see how its abstraction level makes it harder to understand. It's not a lot of code, and something like it can also be copied as well.

> So I think I'm reverting back to my older plan -- a FileSource trait that requires a method that returns impl Iter<Item = (PathLikeType, &[u8])>. And it would also provide documentation_override(path, content) -> Option<bool>, generated_override, etc. methods that would return None by default.

Whatever it's going to be, please assure that the abstraction is truly zero-cost *or* makes everything *even* faster :). These 1.8s for WebKit are more important to me than I dare to admit ;).

**Comment by spenserblack** [comment]
Off-topic, but while I have your attention: turns out gix makes it easier to run the crate in a docker container, too (#97). Just another selling point for gix :wink:

**Comment by Byron** [comment]
Actually this could be bad, but in this case, it's indeed a good thing. `gix` has a different way of dealing with untrusted repositories. Instead of downright refusing to do anything on a repository owned by somebody else it will simply be extra-careful with anything executable by skipping configuration that comes from untrusted configuration files. Reading the repo itself is assumed to be safe as `gix` trusts in its fuzzed parsers.

It's understandable that `git2` and `git` implemented it this way, it's easier, but `gix` was in a unique position here and I think was all worth it.

For example, when extracting git filter driver configuration [it will filter by trust](https://github.com/Byron/gitoxide/blob/072ee32f693a31161cd6a843da6582d13efbb20b/gix/src/filter.rs#L207) to skip all drivers configured by somebody else.

**Comment by spenserblack** [comment]
I was thinking less about trust levels and more about how git2 was producing segfaults :laughing:

As far as trust goes, IMO as long as the gix library makes it clear in docs how repository trust is handled, and perhaps provides scenarios for when one might want more or less trust, things are alright.

**Comment by Byron** [comment]
> As far as trust goes, IMO as long as the gix library makes it clear in docs how repository trust is handled, and perhaps provides scenarios for when one might want more or less trust, things are alright.

I don't think I make it clear anyway, a good point, I have [made an improvement](https://github.com/Byron/gitoxide/commit/e022096aa495f55a05f83860243f49552be501f7).

**Comment by spenserblack** [comment]
@Byron quick question about gix:

Are A and B roughly equivalent, except for the fact that B doesn't take ownership?
```rust
let state: gix::index::State;
// ...

// A
let (entries, path_storage) = state.into_entries();

// B
let (entries, path_storage) = (state.entries(), state.path_backing());
```

Without jumping into the source code, I wasn't 100% sure since `Vec<u8, Global>` is the return type in the docs instead of the `PathStorage` alias.

******

Refactoring this code has given me a new appreciation for gix! I'm embarrassed to admit that I merged #157 without fully understanding how the code worked, but now I see all the caching and performance optimizations :rocket:

**Comment by Byron** [comment]
Yes, that's correct. I don't know if it's a feature or a bug that type definitions aren't necessarily showing up in rustdoc.
